### PR TITLE
Fix backwards compatibility tests in 2.x

### LINKED ISSUE #13425: Backwards compatibility tests broken for 2.x
#13130 uncovered that the backwards compatibility tests don't really work in 2.x. There are a few issues:
1. We don't have a release of 2.x so the best we can do to test 2.x is to use 2.0.0-beta1. But `Version#minimumCompatibilityVersion` seems to reject that. Its simple enough to just make it accept the beta for now.
2. `CompositeTestCluster` seems to be a bit broken.
3. The backwards section in TESTING.asciidoc is super out of date.

### LINKED ISSUE #13484: Backwards compatibility tests should be run with the security manager
Right now the backwards compatibility tests disable the security manager entirely or the external nodes won't run. We should fix that.

### PR DESCRIPTION
**Pull request by nik9000**
These tests have gone stale because 2.0 hasn't needed to be protocol compatible with anything. But 2.0.1 and 2.1.0 will have to be backwards compatible with 2.0.0. Since we don't have 2.0.0 released but we still want to work on features for 2.1.0 that have to be backwards compatible with 2.0.0 we should use 2.0.0-beta1 as a stand in for the eventual 2.0.0 release.

Closes #13425
Closes #13484

### CODE REVIEW

**Comment by nik9000** [comment]
This is very much a work in progress. It'll take a bit more banging around to fix the remaining BWC tests.

File: TESTING.asciidoc [hunk 1]
```diff
@@ -221,30 +221,26 @@ mvn test -Dtests.jvm.argline="-XX:HeapDumpPath=/path/to/heapdumps"
...
-Note that backwards tests must be run with security manager disabled.
-If the elasticsearch release is placed under `./backwards/elasticsearch-x.y.z` the path
-can be omitted:
+To setup the bwc test environment execute the following steps (provided you are
+already in your elasticsearch clone):
 
 ---------------------------------------------------------------------------
-mvn test -Dtests.filter="@backwards" -Dtests.bwc.version=x.y.z -Dtests.security.manager=false
+$ mkdir -p core/backwards && pushd core/backwards
+$ curl -O https://download.elasticsearch.org/elasticsearch/release/org/elasticsearch/distribution/tar/elasticsearch/2.0.0-beta1/elasticsearch-2.0.0-beta1.tar.gz
```
Comment (rjernst) [inline TESTING.asciidoc]: Why not use the provided dev-tools script for this? get-bwc-version.py

Comment (nik9000) [inline TESTING.asciidoc]: Because I forgot about it. I'll update.

File: core/src/test/java/org/elasticsearch/search/functionscore/FunctionScoreBackwardCompatibilityIT.java [hunk 2]
```diff
@@ -46,7 +46,6 @@
     /**
      * Simple upgrade test for function score
      */
-    @Test
```
Comment (rjernst) [inline core/src/test/java/org/elasticsearch/search/functionscore/FunctionScoreBackwardCompatibilityIT.java]: You are a man among men!

File: core/src/test/java/org/elasticsearch/test/ExternalNode.java [hunk 3]
```diff
@@ -130,6 +132,10 @@ synchronized void startInternal(Client client, Settings settings, String nodeNam
 
         params.add("-Des.path.home=" + PathUtils.get(".").toAbsolutePath());
         params.add("-Des.path.conf=" + path + "/config");
+        // Security manager is getting in the way. We should turn it back on though.
```
Comment (rjernst) [inline core/src/test/java/org/elasticsearch/test/ExternalNode.java]: Can you open an issue for this?

File: core/src/test/java/org/elasticsearch/test/ExternalNode.java [hunk 4]
```diff
@@ -199,6 +212,7 @@ synchronized Client getClient() {
             Settings clientSettings = settingsBuilder().put(externalNodeSettings)
                     .put("client.transport.nodes_sampler_interval", "1s")
                     .put("name", "transport_client_" + nodeInfo.getNode().name())
+                    .put("path.home", PathUtils.get(".").toAbsolutePath())
```
Comment (rjernst) [inline core/src/test/java/org/elasticsearch/test/ExternalNode.java]: Hrm, why is this cwd?

**Comment by rjernst** [comment]
Thanks @nik9000! I left a couple comments. The changes look fine, I think.

File: core/src/test/java/org/elasticsearch/test/ExternalNode.java (hunk 3 above)
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/test/ExternalNode.java]: https://github.com/elastic/elasticsearch/issues/13484

File: core/src/test/java/org/elasticsearch/test/ExternalNode.java (hunk 4 above)
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/test/ExternalNode.java]: Its the CWD of the external node which I picked because it worked but without any strong reasoning. If you have a better idea I'll certainly use it!

File: core/src/test/java/org/elasticsearch/search/functionscore/FunctionScoreBackwardCompatibilityIT.java [hunk 5]
```diff
@@ -62,7 +61,7 @@ public void testSimpleFunctionScoreParsingWorks() throws IOException, ExecutionE
                         .endObject()
                         .endObject()
                         .endObject()
-                        .endObject()));
+                    .endObject()).setSettings("number_of_shards", 1));
```
Comment (brwe) [inline core/src/test/java/org/elasticsearch/search/functionscore/FunctionScoreBackwardCompatibilityIT.java]: why is this needed?

Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/functionscore/FunctionScoreBackwardCompatibilityIT.java]: Sorry, that was burried in a commit message:
`FunctionScoreBackwardCompatibilityIT` relied on tf/idf information but didn't
set the number of shards to 1 or use dfs_query_then_fetch. Now it uses one
shard.

I'll add it to a comment here if you approve.

**Comment by nik9000** [comment]
I'm dropping 2.0.0 from this because its not strictly required for 2.0.0 - its needed for 2.0.1 and 2.1.0. We'll get this soon but I don't want it to be a blocker.

**Comment by nik9000** [comment]
Ok - I've finally got this ready for review. It doesn't fix all the tests but I was sinking hour and hours and hours per busted test and I didn't want each one to block getting the rest fixed.

I did squash and rebase this during the review and for that I'm sorry - it was getting pretty sprawling and had some twists and turns that ended up not being good ways to fix the tests. But its really really ready for review now.

**Comment by nik9000** [comment]
Added 3.0 label. This should be merged into 2.0, 2.x, and master branches even though we don't expect master to be backwards compatible with anything.

**Comment by nik9000** [comment]
Ping to @rjernst and probably @dakrone to review.

I've disabled custom data paths for these tests for now because they don't seem to be working properly - probably because the data paths are relative to the path.home and the upgraded nodes have a different path.home than the backwards nodes.

File: core/src/test/java/org/elasticsearch/test/ESBackcompatTestCase.java [hunk 6]
```diff
@@ -141,6 +143,18 @@ private static Path backwardsCompatibilityPath() {
...
 
+    @Override
+    public Settings indexSettings() {
+        // IndexMetaData.SETTING_DATA_PATH looks like it breaks backwards compatibility tests
+        Settings settings = super.indexSettings();
+        if (settings.get(IndexMetaData.SETTING_DATA_PATH) == null) {
+            return settings;
+        }
+        Settings.Builder builder = Settings.settingsBuilder().put(settings);
+        builder.remove(IndexMetaData.SETTING_DATA_PATH);
```
Comment (dakrone) [inline core/src/test/java/org/elasticsearch/test/ESBackcompatTestCase.java]: I think you should log something here to make it apparent to anyone reading that the setting was removed? Maybe at `warn` level?

Comment (nik9000) [inline core/src/test/java/org/elasticsearch/test/ESBackcompatTestCase.java]: Done

File: TESTING.asciidoc [hunk 7]
```diff
@@ -221,30 +221,40 @@ mvn test -Dtests.jvm.argline="-XX:HeapDumpPath=/path/to/heapdumps"
...
 requires a release version of elasticsearch to be present on the test system.
-To run backwards compatibilty tests untar or unzip a release and run the tests
-with the following command:
 
+Until there is a GA release in the 2.0 line follow these temporary instructions
+to setup the bwc test environment execute the following steps:
+
+
+1. Go to
+https://oss.sonatype.org/content/repositories/snapshots/org/elasticsearch/distribution/tar/elasticsearch/2.0.0-beta2-SNAPSHOT/
```
Comment (javanna) [inline TESTING.asciidoc]: we can use rc1 now, but I certainly hope that by the time this PR gets, GA is released

**Comment by nik9000** [comment]
Quick update for anyone following along at home: as of the last patch the BWC tests that are not `@AwaitsFix` are now passing reasonably consistently. All but `UnicastBackwardsCompatibilityIT` pass all the times I've tried them. I'm working on that last one.

After that I think we have two options:
1. Just merge this.
2. I keep working on un `@AwaitsFix`ing things until 2.0 is released and then I switch this to use 2.0 and roll back the `Version` hack I added way back at the beginning of this process and _then_ we merge it.

File: TESTING.asciidoc [hunk 8]
```diff
@@ -312,6 +284,46 @@ You can also skip this by using the "dev" profile:
...
+Its fairly easy to fetch one though:
+
+---------------------------------------------------------------------------
+python3 dev-tools/get-bwc-version.py 2.0.0-rc1
+---------------------------------------------------------------------------
+
+Then you can run the backwards compatibility tests:
+
+---------------------------------------------------------------------------
+mvn verify -pl core -Dskip.unit.tests -Pdev -Dtests.filter="@backwards and not @awaitsfix" -Dtests.bwc.version=2.0.0-rc1 -Dtests.bwc.path="$(pwd)/backwards" -Dtests.security.manager=false -Dtests.seccomp=false -Dtests.jvms=1
```
Comment (javanna) [inline TESTING.asciidoc]: are -Pdev and -Dtests.jvms=1 required?

File: TESTING.asciidoc [hunk 9]
```diff
@@ -312,6 +284,46 @@ You can also skip this by using the "dev" profile:
...
+just the backwards compatibility tests that aren't marked as broken.
+* `-Dtests.bwc.version=2.0.0-rc1`: Specify the older version to run against.
+* `-Dtests.bwc.path="$(pwd)/backwards"`: Specify the directory where the older
+version lives.
+* `-Dtests.security.manager=false`: Turn off the security manager because it
+gets in the way of the backwards compatibility tests. We want to improve the
+tests so that this isn't required.
+* `-Dtests.seccomp=false`: Seccomp stops you from forking and the backwards
+compatiblity tests fork the backwards nodes. So we can't have this at all!
+* `-Dtests.jvms=1`: The backwards compatibility tests communicate over the
```
Comment (javanna) [inline TESTING.asciidoc]: I missed there was an explanation on this, sorry.  I don't remember this being required before, maybe I am wrong. Is it a new requirement due to recent changes? I thought that like with integration test, each jvm would be isolated and it would be possible to distribute tests over them (not that it's crazy important at this stage, just curious)

File: core/src/main/java/org/elasticsearch/Version.java [hunk 10]
```diff
@@ -625,6 +625,12 @@ public boolean onOrBefore(Version version) {
      * is a beta or RC release then the version itself is returned.
      */
     public Version minimumCompatibilityVersion() {
+        if (major == 2) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/Version.java]: add a nocommit just to make sure we don't forget? ;)

File: TESTING.asciidoc (hunk 9 above)
Comment (nik9000) [inline TESTING.asciidoc]: I haven't tried it with multiple jvms in a month or so but my memory is that it failed spectacularly. I'm happy to give it another shot/try to get it working. The failures were so impressive when I started on this that I didn't want anything else complicating it though.

File: core/src/main/java/org/elasticsearch/Version.java (hunk 10 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/Version.java]: norelease?

File: core/src/test/java/org/elasticsearch/bwcompat/BasicBackwardsCompatibilityIT.java [hunk 11]
```diff
@@ -371,16 +380,13 @@ public void assertVersionCreated(Version version, String... indices) {
...
     public void testUnsupportedFeatures() throws IOException {
         XContentBuilder mapping = XContentBuilder.builder(JsonXContent.jsonXContent)
                 .startObject()
                     .startObject("type")
                         .startObject(FieldNamesFieldMapper.NAME)
-                           // by setting randomly index to no we also test the pre-1.3 behavior
-                            .field("index", randomFrom("no", "not_analyzed"))
-                            .field("store", randomFrom("no", "yes"))
+                            // by setting randomly enabled to no we also test the pre-1.3 behavior
+                            .field("enabled", randomFrom("no", "yes"))
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/bwcompat/BasicBackwardsCompatibilityIT.java]: does enabled replace index and store?

File: core/src/test/java/org/elasticsearch/bwcompat/BasicBackwardsCompatibilityIT.java [hunk 12]
```diff
@@ -602,10 +601,11 @@ public void testExplain() {
         assertThat(response.isMatch(), equalTo(true));
         assertThat(response.getExplanation(), notNullValue());
         assertThat(response.getExplanation().isMatch(), equalTo(true));
-        assertThat(response.getExplanation().getDetails().length, equalTo(1));
+        System.err.println(response.getExplanation().getDetails().length);
+        System.err.println(java.util.Arrays.deepToString(response.getExplanation().getDetails()));
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/bwcompat/BasicBackwardsCompatibilityIT.java]: leftovers

File: core/src/test/java/org/elasticsearch/bwcompat/ClusterStateBackwardsCompatIT.java [hunk 13]
```diff
@@ -102,6 +100,7 @@ private NodesInfoResponse clusterNodes() {
 
     private TransportClient newTransportClient() {
         Settings settings = Settings.settingsBuilder().put("client.transport.ignore_cluster_name", true)
+                .put("path.home", PathUtils.get(".").toAbsolutePath())
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/bwcompat/ClusterStateBackwardsCompatIT.java]: is `PathUtils.get(".").toAbsolutePath()` correct? I see that we do `.put("path.home", createTempDir().toAbsolutePath())` in a lot of existing tests....

File: core/src/test/java/org/elasticsearch/snapshots/SnapshotBackwardsCompatibilityIT.java [hunk 14]
```diff
@@ -220,26 +218,26 @@ public void testSnapshotMoreThanOnce() throws ExecutionException, InterruptedExc
...
+        assertThat(client().admin().cluster().prepareGetSnapshots("test-repo").setSnapshots("test-1").get().getSnapshots().get(0).state(), equalTo(SnapshotState.SUCCESS));
         {
-            SnapshotStatus snapshotStatus = client.admin().cluster().prepareSnapshotStatus("test-repo").setSnapshots("test-1").get().getSnapshots().get(0);
+            SnapshotStatus snapshotStatus = client().admin().cluster().prepareSnapshotStatus("test-repo").setSnapshots("test-1").get().getSnapshots().get(0);
             List<SnapshotIndexShardStatus> shards = snapshotStatus.getShards();
             for (SnapshotIndexShardStatus status : shards) {
-
-                assertThat(status.getStats().getProcessedFiles(), equalTo(1)); // we flush before the snapshot such that we have to process the segments_N files
+                // Nothing has changed since the last snapshot so there is nothing to save.
+                assertThat(status.getStats().getProcessedFiles(), equalTo(0));
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/snapshots/SnapshotBackwardsCompatibilityIT.java]: sorry for asking lots of questions... why does this become 0 if it was 1 before?

File: core/src/test/java/org/elasticsearch/test/ESBackcompatTestCase.java [hunk 15]
```diff
@@ -141,6 +143,20 @@ private static Path backwardsCompatibilityPath() {
         return builder;
     }
 
+    @Override
+    public Settings indexSettings() {
+        // IndexMetaData.SETTING_DATA_PATH looks like it breaks backwards compatibility tests
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/test/ESBackcompatTestCase.java]: wonder whether it's ok that it breaks tests, or whether we should dig... I think it shouldn't break things in theory...

**Comment by javanna** [comment]
I had a quick look and left some comments. Also ran tests on my machine, UpgradeIT failed with `java.nio.file.NoSuchFileException: /Users/lucacavanna/dev/elasticsearch/core/core/target/J0/./temp/bwc/external_1/data`, maybe something is wrong on my machine? Also added 2.0.1 label as this will need to go to 2.0 branch too.

File: core/src/test/java/org/elasticsearch/bwcompat/BasicBackwardsCompatibilityIT.java (hunk 11 above)
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/bwcompat/BasicBackwardsCompatibilityIT.java]: This is that internal field names field used for the _missing query. It isn't allowed to have `index` or `store` any more. But `enabled` replaces `index` in this case. `store` was weird, iirc. We didn't need it stored to use it so this was just twiddling that bit for the sake of it I think. Because it was an option.

File: core/src/test/java/org/elasticsearch/test/ESBackcompatTestCase.java (hunk 15 above)
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/test/ESBackcompatTestCase.java]: I think this is one of the things that I'd like to fix in https://github.com/elastic/elasticsearch/issues/13522

**Comment by nik9000** [comment]
> Also added 2.0.1 label as this will need to go to 2.0 branch too.

Awesome.

> UpgradeIT failed

Booo! It didn't fail for me but let me try again.

File: core/src/test/java/org/elasticsearch/bwcompat/ClusterStateBackwardsCompatIT.java (hunk 13 above)
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/bwcompat/ClusterStateBackwardsCompatIT.java]: Hmm - the transport client doesn't really _use_ `path.home` so it can be anything. I can use that instead.

File: core/src/test/java/org/elasticsearch/snapshots/SnapshotBackwardsCompatibilityIT.java (hunk 14 above)
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/snapshots/SnapshotBackwardsCompatibilityIT.java]: This is one that I remember starting to debug - I'll just revert it.

File: TESTING.asciidoc (hunk 8 above)
Comment (nik9000) [inline TESTING.asciidoc]: -Pdev isn't really required, but it does speed up the process.

File: core/src/main/java/org/elasticsearch/Version.java (hunk 10 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/Version.java]: Done.

File: core/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java [hunk 16]
```diff
@@ -93,12 +93,14 @@ public static void initializeNatives(Path tmpFile, boolean mlockAll, boolean sec
...
             }
         }
-        
+
         // enable secure computing mode
         if (seccomp) {
             Natives.trySeccomp(tmpFile);
+        } else {
+            logger.warn("Not enabling seccomp because it was explicitly disabled");
         }
```
Comment (rmuir) [inline core/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java]: Can we remove the whitespace changes and the warning? I would also not like to document this in TESTING.md or anywhere else. It defeats the whole point of it. I also don't think we should support allowing tests to do this for any reason, but I know I stand alone.

File: pom.xml [hunk 17]
```diff
@@ -110,6 +110,7 @@
         <tests.security.manager>true</tests.security.manager>
         <tests.compatibility></tests.compatibility>
         <tests.ifNoTests>fail</tests.ifNoTests>
+        <tests.seccomp>true</tests.seccomp>
```
Comment (rmuir) [inline pom.xml]: Can we add an xml comment "not a supported parameter".

File: pom.xml [hunk 18]
```diff
@@ -670,6 +671,8 @@
                             <es.node.local>${env.ES_TEST_LOCAL}</es.node.local>
                             <es.node.mode>${es.node.mode}</es.node.mode>
                             <es.logger.level>${es.logger.level}</es.logger.level>
+                            <!-- Backwards compatibility tests need to disable seccomp -->
+                            <tests.seccomp>${tests.seccomp}</tests.seccomp>
                             <tests.security.manager>${tests.security.manager}</tests.security.manager>
```
Comment (rmuir) [inline pom.xml]: here too.

File: core/src/test/java/org/elasticsearch/bootstrap/BootstrapForTesting.java [hunk 19]
```diff
@@ -68,11 +68,12 @@
         }
 
         // just like bootstrap, initialize natives, then SM
-        Bootstrap.initializeNatives(javaTmpDir, true, true, true);
+        boolean seccomp = Boolean.parseBoolean(System.getProperty("tests.seccomp", "true"));
```
Comment (rmuir) [inline core/src/test/java/org/elasticsearch/bootstrap/BootstrapForTesting.java]: can this use a non-lenient boolean parser? it shoudl be hard, like really hard, to turn this off. not the other way around.

**Comment by rmuir** [comment]
I'm really really concerned about the execution here. I do not think it should be allowed, even in tests.

If these tests must really run this way (which i dont believe), we should start up some other background test listening on a socket or similar responsible for launching them, instead of allowing execution from the ES environment. (i already see this "setting a precendent" for more leniency).

File: core/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java (hunk 16 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java]: > Can we remove the whitespace changes and the warning?

I can.

> I would also not like to document this in TESTING.md or anywhere else. It defeats the whole point of it. I also don't think we should support allowing tests to do this for any reason, but I know I stand alone.

I'm happy to spend the next few days working on getting it running under seccomp.

Comment (rmuir) [inline core/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java]: I just think we can launch a little perl script in pre-integration-test, kill it in post-integration-test, that sits on a socket and starts up nodes. Its a perl script for testing. Then the rest of the stuff can run without us having to defeat every security mechanism we are trying to add...

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java]: Yeah - it looks like some kind of remote executor helper is required for this. I feel like it'd be easier to yank some of the code we already use to start the bwc nodes into a java process started into pre-integration-test than to write it in perl and fight with windows.

Comment (rmuir) [inline core/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java]: Well, i dont really think it is. I guess i don't understand why we are running backwards tests this way, versus doing real integration tests with the rest suite and similar (like we do in qa). All the stuff needed for that is published as maven artifacts. One of these approaches is a "fantasy world" that isn't quite realistic, it requires manual setup and is difficult to run, and flaky as far as jenkins goes. the other is automatic, realistic and reliable... what am I missing?

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java]: I agree that the dependency resolution is funky here. I think we'd be better off with dependency resolution handled by maven/gradle/whatever then how it works now. I suspect the reason the resolution isn't done that way now is because these are left over from single module days.

I think giving the test process some degree of control of the bwc nodes is useful though - it wants to be able to shut them down and spin up new nodes to replace them to simulate rolling upgrades which seems pretty fair.

The way I see this we need these tests working as soon as we want to make 2.0.1 so I don't know that we want to do major surgery to them. I'm all for figuring a way to get them working under seccomp and I'll work on it. With the goal of eventually moving them into the qa project with real resolution. And no elasticsearch nodes running in process with the tests.

**Comment by s1monw** [comment]
> I'm really really concerned about the execution here. I do not think it should be allowed, even in tests.

the reason why we run this way is a historic one. This is how I added it back in 1.2 when we had no BWC tests at all. From the todays perspective I think we should move away from how it's done. The question is if it makes sense to make the test work again and then work from there to cut over to a new system that is controlled from the outside ie. basic tests that replace the once that we have? 
To me what is most important are 2 kinds of tests:

 - upgrade tests where we upgrade a cluster from one version to another either as _rolling_ or _full cluster_ restart
 - running simple tests against a mixed version cluster to ensure our wire protocols are working and we can operate in such a scenario

both can relatively simple be done without starting / stopping nodes from within ES and maybe thats a good start? I still wonder if we should make the ones we have work first and then go from there?

**Comment by nik9000** [comment]
> I still wonder if we should make the ones we have work first and then go from there?

This patch as it stands gets them working in the same way they worked before.

I'm going to work on a little daemon we can start pre-integration-test that can shim the start/stop of `ExternalNode`s. We could cut over to using that and keep the tests as they stand and reenable seccomp. I'd be much happier with the PR if we can do that.

There are lots of great things we could do here - move to the qa module and get proper maven dependency resolution so they run on every run - invoke current version of elasticsearch as an `ExternalNode` so its less mock-ed.

**Comment by nik9000** [comment]
OK - I've build a service that can be forked pre-integration-test to run these the external nodes. Right now it forks in core's build only if `tests.bwc.path` is set. I'd love to move all of these tests to qa and turn on proper dependency resolution so we don't need `tests.bwc.path`.

I'm in no way proud of the service's implementation - it gets the job done but I'm sure I've done tons of non normal stuff so feel free to complain about it.

**Comment by nik9000** [comment]
Another update: the tests now pass with the security manager enabled so I've removed the recommendation to disable it. The external nodes and the internal nodes run under the security manager. Only ExternalNodeService runs without it and it runs in the maven JVM, at least it does the way I've got it rigged right now.

**Comment by nik9000** [comment]
Yet another update: the tests now no longer need the external setup and download step, they resolve their dependencies using maven.

File: core/src/test/java/org/elasticsearch/test/ESIntegTestCase.java [hunk 20]
```diff
@@ -271,6 +277,27 @@
...
+    /*
+     * These are a few backwards compatibility test constants that are declared
+     * here so they can be used in both ESBackcompatTestCase and
+     * ReproduceInfoPrinter.
+     */
+    /**
+     * Key used to set the path for the elasticsearch executable used to run backwards compatibility tests from
+     * via the commandline -D{@value #TESTS_BACKWARDS_COMPATIBILITY}
+     */
+    public static final String TESTS_BACKWARDS_COMPATIBILITY = "tests.bwc";
```
Comment (rjernst) [inline core/src/test/java/org/elasticsearch/test/ESIntegTestCase.java]: Hrm, can we rather allow customizing from repro info from subclasses of ESIntegTestCase? I don't like having to add everything here...

Comment (nik9000) [inline core/src/test/java/org/elasticsearch/test/ESIntegTestCase.java]: Yeah - I agree - I think this is nasty. I did it without a whole lot of thought trying to get the move working.

File: qa/backwards/shared/src/test/java/org/elasticsearch/test/ESBackcompatTestCase.java [hunk 21]
```diff
@@ -212,17 +196,14 @@ private Settings addLoggerSettings(Settings externalNodesSettings) {
...
+    protected int minExternalNodes() {
+        return 1;
+    }
 
     protected int maxExternalNodes() {
         return 2;
     }
 
-    @Override
-    protected int maximumNumberOfReplicas() {
```
Comment (rjernst) [inline qa/backwards/shared/src/test/java/org/elasticsearch/test/ESBackcompatTestCase.java]: Not sure if this would work, but could you keep this override but make it relative to the number of nodes on latest version, essentially removing the need for hte new "num lucene latest version" method? That is a very awkward parameter for a general integ test that is really specific to backcompat tests...

**Comment by rjernst** [comment]
LGTM, let's get this into 2.x/2.1. It won't work on master until we port to gradle, but maybe by then we can re-evaluate how to do bwc tests to simplify this (and use all real nodes instead of some mock and some external).

File: qa/backwards/shared/src/test/java/org/elasticsearch/test/ESBackcompatTestCase.java (hunk 21 above)
Comment (nik9000) [inline qa/backwards/shared/src/test/java/org/elasticsearch/test/ESBackcompatTestCase.java]: Yeah - it really is. I imagine I could move maximumNumberOfReplicas into TestCluster and that'd be nicer.

**Comment by nik9000** [comment]
Squashed and rebased.

**Comment by nik9000** [comment]
Merged to 2.x. Cherry-picking to 2.1 now.

**Comment by nik9000** [comment]
And merged into 2.1: 10315d8167de96fc0b2b992b5407a8518b379a94
